ddl/db_test.go

```diff
@@ -2450,41 +2453,166 @@ func (s *testDBSuite) TestAlterTableAddPartition(c *C) {
+	s.tk.MustExec(`create table employees (
+	id int not null,
+	hired int not null
+	)
+	partition by range( hired ) (
+		partition p1 values less than (1991),
+		partition p2 values less than (1996),
+		partition p3 values less than (2001)
+	);`)
+
+	s.tk.MustExec("alter table employees drop partition p3;")
+	ctx := s.tk.Se.(sessionctx.Context)
+	is := domain.GetDomain(ctx).InfoSchema()
+	tbl, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("employees"))
+	c.Assert(err, IsNil)
+	c.Assert(tbl.Meta().GetPartitionInfo(), NotNil)
+	part := tbl.Meta().Partition
+	c.Assert(part.Type, Equals, model.PartitionTypeRange)
+	c.Assert(part.Expr, Equals, "`hired`")
+
+	c.Assert(part.Definitions, HasLen, 2)
+	c.Assert(part.Definitions[0].LessThan[0], Equals, "1991")
+	c.Assert(part.Definitions[0].Name, Equals, model.NewCIStr("p1"))
+	c.Assert(part.Definitions[1].LessThan[0], Equals, "1996")
+	c.Assert(part.Definitions[1].Name, Equals, model.NewCIStr("p2"))
+
+	s.tk.MustExec("drop table if exists table1;")
+	s.tk.MustExec("create table table1 (a int);")
+	sql1 := "alter table table1 drop partition p10;"
+	s.testErrorCode(c, sql1, tmysql.ErrPartitionMgmtOnNonpartitioned)
+
+	s.tk.MustExec("drop table if exists table2;")
+	s.tk.MustExec(`create table table2 (
+	id int not null,
+	hired date not null
+	)
+	partition by range( year(hired) ) (
+		partition p1 values less than (1991),
+		partition p2 values less than (1996),
+		partition p3 values less than (2001)
+	);`)
+	sql2 := "alter table table2 drop partition p10;"
+	s.testErrorCode(c, sql2, tmysql.ErrDropPartitionNonExistent)
+
+	s.tk.MustExec("drop table if exists table3;")
+	s.tk.MustExec(`create table table3 (
+	id int not null
+	)
+	partition by range( id ) (
+		partition p1 values less than (1991)
+	);`)
+	sql3 := "alter table table3 drop partition p1;"
+	s.testErrorCode(c, sql3, tmysql.ErrDropLastPartition)
+
+	s.tk.MustExec("drop table if exists table4;")
+	s.tk.MustExec(`create table table4 (
+	id int not null
+	)
+	partition by range( id ) (
+		partition p1 values less than (10),
+		partition p2 values less than (20),
+		partition p3 values less than MAXVALUE
+	);`)
+
+	s.tk.MustExec("alter table table4 drop partition p2;")
+	is = domain.GetDomain(ctx).InfoSchema()
+	tbl, err = is.TableByName(model.NewCIStr("test"), model.NewCIStr("table4"))
+	c.Assert(err, IsNil)
+	c.Assert(tbl.Meta().GetPartitionInfo(), NotNil)
+	part = tbl.Meta().Partition
+	c.Assert(part.Type, Equals, model.PartitionTypeRange)
+	c.Assert(part.Expr, Equals, "`id`")
+
+	c.Assert(part.Definitions, HasLen, 2)
+	c.Assert(part.Definitions[0].LessThan[0], Equals, "10")
+	c.Assert(part.Definitions[0].Name, Equals, model.NewCIStr("p1"))
+	c.Assert(part.Definitions[1].LessThan[0], Equals, "MAXVALUE")
+	c.Assert(part.Definitions[1].Name, Equals, model.NewCIStr("p3"))
+
+	s.tk.MustExec("drop table if exists tr;")
+	s.tk.MustExec(` create table tr(
+		id int, name varchar(50), 
+		purchased date
```

Comment:
Could we add a key for `purchased`? And check the index rows.

Reply:
@zimulala Need to wait for this `PR` #6814 to complete before adding it.